C++ PID

mxmllnknz opened the pull request.

mxmllnknz committed:
Basic C++ PID by Tanvika Dasari

sethgi commented on controls/include/controls/PID.h:
```diff
@@ -0,0 +1,38 @@
+// Created by tanvi on 10/30/2020.
+//
+
+#ifndef CONTROLS_PID_H
```
Replace these old-style include guards with a single `#pragma once`

This replaces the entire include guard, including the `#endif` at the end of the file.

sethgi commented on controls/include/controls/PID.h:
```diff
@@ -0,0 +1,38 @@
+
+public:
+    PID(double kP, double kI, double kD);
+    PID(double kP, double kI, double kD, bool angle);
```
Maybe rename angle to isAngle or something to indicate why, and drop a comment indicating what the argument is for

sethgi commented on controls/include/controls/PID.h:
```diff
@@ -0,0 +1,38 @@
+
+    void tune(double kP, double kI, double kD);
+
+    double getkI();
```
I would camelcase these for style. So, instead of `getkI do `getKI`.

sethgi commented on controls/include/controls/six_deg_PID.h:
```diff
@@ -0,0 +1,44 @@
+// Created by tanvi on 11/21/2020.
+//
+
+#ifndef CONTROLS_SIX_DEG_PID_H
```
Again, use `#pragma once` instead of an include guard here

sethgi commented on controls/include/controls/six_deg_PID.h:
```diff
@@ -0,0 +1,44 @@
+#include "PID.h"
+#include <vector>
+
+class six_deg_PID
```
Class names should be all caps, no underscores. So, `SixDegPID`. Also, I would generally avoid abbreviations as it is not that much harder to type out (esp. with tab complete) and much easier to read. So, I would rename this class to `SixDegreePID`

sethgi commented on controls/include/controls/six_deg_PID.h:
```diff
@@ -0,0 +1,44 @@
+
+    void tuneAngle(std::vector<double> kP, std::vector<double> kI, std::vector<double> kD);
+
+    void tunePosition(std::vector<double> kP, std::vector<double> kI, std::vector<double> kD);
```
In my opinion this interface may be tricky to work with. In general, we tune one PID loop at a time. So, it may be a bit inconvenient to make these vectors of coefficients. I would recommend either having six functions, one per controller, or just one which takes the loop to tune as an argument (probably passed by reference). So the options would be:

```
void tuneSurge(double kP, double kI, double kD);
void tuneSway(double kP, double kI, double kD);
...
```

OR 

`void tuneController(double kP, double kI, double kD, const PID& controller)`

sethgi commented on controls/include/controls/six_deg_PID.h:
```diff
@@ -0,0 +1,44 @@
+    PID pitch_ = PID();
+    PID yaw_ = PID();
+
+    PID x_ = PID();
```
I would rename x,y,z to surge,sway,heave. These are the terms commonly used in marine settings and underwater robotics. The reason it's helpful is x,y, and z may be defined differently on different vehicles. For instance, sometimes x is forward and sometimes y is forward. But, with surge, sway, heave, surge is always forward, sway is always lateral, and heave is always vertical.

sethgi commented on controls/include/controls/six_deg_PID.h:
```diff
@@ -0,0 +1,44 @@
+
+    void tunePosition(std::vector<double> kP, std::vector<double> kI, std::vector<double> kD);
+
+    std::vector<double> update(std::vector<double> setPoint, std::vector<double> plantState, double dT);
```
A comment here indicating the expected format of the vectors would be nice.

sethgi commented on controls/src/PID.cc:
```diff
@@ -0,0 +1,75 @@
+            error = otherError;
+        }
+
+        //std::cout<<"error"<<error<<' '<<goal<<' '<<current<<std::endl;
```
delete before merging

sethgi commented on controls/src/PID.cc:
```diff
@@ -0,0 +1,75 @@
+    if(angle_)
+    {
+
+        double otherError = (360 - error)*-1;
```
variable naming

sethgi commented on controls/src/PID.cc:
```diff
@@ -0,0 +1,75 @@
+        //std::cout<<"error"<<error<<' '<<goal<<' '<<current<<std::endl;
+    }
+    integralError_ += error * deltaT;
+    double derivativeError = (error + previousError_) / 2;
```
Shouldn't this be `(error - previousError)/deltaT`?

sethgi commented on controls/src/PID.cc:
```diff
@@ -0,0 +1,75 @@
+
+        //std::cout<<"error"<<error<<' '<<goal<<' '<<current<<std::endl;
+    }
+    integralError_ += error * deltaT;
```
One important thing with PID controllers is to have what's called a windup limit on the integral error. Basically, you don't wan't the integral term to freak out if you're away from the setpoint for too long. So, I would create a member variable called `windupLimit` which defaults to 100. Then, add a check here something like:

```
integralError_ = min(integralError_, windupLimit_);
integrallError_ =  max(integralError_, -1*windupLimit);
```

tanvikad committed:
Updated PID code based on review feedback

sethgi committed:
Update PID.h

sethgi commented on controls/include/controls/SixDegreePID.h:
```diff
@@ -0,0 +1,39 @@
+class SixDegreePID
+{
+private:
+    PID roll_ = PID();
```
PID surge_, sway_, heave_, roll_, pitch_, yaw_;

sethgi commented on controls/include/controls/SixDegreePID.h:
```diff
@@ -0,0 +1,39 @@
+    PID surge_ = PID();
+    PID sway_ = PID();
+    PID heave_ = PID();
+
```
struct Axes
{
  double surge_, sway_, heave_, roll_, pitch_, yaw_;
}

sethgi commented on controls/src/PID.cc:
```diff
@@ -0,0 +1,91 @@
+    integralError_ = std::min(integralError_, windupLimit_);
+    integralError_ =  std::max(integralError_, -1*windupLimit_);
+
+    double derivativeError = (error - previousError_) / 2;
```
/deltaT

sethgi commented on controls/src/SixDegreePID.cc:
```diff
@@ -0,0 +1,70 @@
+
+SixDegreePID::SixDegreePID(std::vector<double> kP, std::vector<double> kI, std::vector<double> kD)
+{
+    roll_ = PID(kP.at(0), kI.at(0), kD.at(0), true);
```
kP[0]

tanvikad committed:
Removed the two six_deg_PID overlapping files

tanvikad committed:
Merge branch 'develop/controls' of https://github.com/MuddSub/MuddSub into develop/controls

sethgi committed:
Integrated into sim (untested)

sethgi committed:
Merge branch 'develop/controls' of https://github.com/MuddSub/MuddSub into develop/controls

sethgi committed:
Added simple tuning interface

sethgi committed:
tuning tested

sethgi committed:
Fixes to dynamics

tanvikad committed:
untested style changes added to PID controller

sethgi committed:
Moved typedefs around

Eli-Pregerson committed:
first prototype of MPC with pseudo model dynamics numbers

Alex-Fay committed:
fixed var documentation, add dynamic matrix 

fixed from after Wednesday night meeting

Eli-Pregerson committed:
Updated control and state dynamics matrices

Alex-Fay committed:
fixed thruster error

push from Sunday's meeting -- need to add VehcicleDynamics() obj and vehicleDynamics.call = Dynamic matrix next week

Alex-Fay committed:
Dynamic Matrix output from ros node

in progress

Alex-Fay committed:
updated A Matrix
